cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/MinTopicLeadersPerBrokerGoal.java

```diff
@@ -0,0 +1,447 @@
+   *
+   * @param action Action to be checked for acceptance.
+   * @param clusterModel The state of the cluster.
+   * @return {@link ActionAcceptance#ACCEPT} if the action is acceptable by this goal,
+   *         {@link ActionAcceptance#REPLICA_REJECT} otherwise.
+   */
+  @Override
+  public ActionAcceptance actionAcceptance(BalancingAction action, ClusterModel clusterModel) {
+    if (!actionAffectsRelevantTopics(action)) {
+      return ACCEPT;
+    }
+    switch (action.balancingAction()) {
+      case LEADERSHIP_MOVEMENT:
+      case INTER_BROKER_REPLICA_MOVEMENT:
+        Replica replicaToBeRemoved = clusterModel.broker(action.sourceBrokerId()).replica(action.topicPartition());
+        return doesLeaderRemoveViolateOptimizedGoal(replicaToBeRemoved) ? REPLICA_REJECT : ACCEPT;
+      case INTER_BROKER_REPLICA_SWAP:
+        Replica srcReplicaToSwap = clusterModel.broker(action.sourceBrokerId()).replica(action.topicPartition());
+        Replica dstReplicaToSwap = clusterModel.broker(action.destinationBrokerId()).replica(action.destinationTopicPartition());
+        return acceptReplicaSwap(srcReplicaToSwap, dstReplicaToSwap);
+      default:
+        throw new IllegalArgumentException("Unsupported balancing action " + action.balancingAction() + " is provided.");
+    }
+  }
+
+  private ActionAcceptance acceptReplicaSwap(Replica srcReplicaToSwap, Replica dstReplicaToSwap) {
+    if (!srcReplicaToSwap.isLeader() && !dstReplicaToSwap.isLeader()) {
+      // Swapping two follower replicas is always accepted
+      return ACCEPT;
+    }
+    String srcTopicName = srcReplicaToSwap.topicPartition().topic();
+    String dstTopicName = dstReplicaToSwap.topicPartition().topic();
+    if (srcReplicaToSwap.isLeader() && dstReplicaToSwap.isLeader() && Objects.equals(srcTopicName, dstTopicName)) {
+      // Swapping two leader replicas of the same topic is always accepted
+      return ACCEPT;
+    }
+    if (doesLeaderRemoveViolateOptimizedGoal(srcReplicaToSwap) || doesLeaderRemoveViolateOptimizedGoal(dstReplicaToSwap)) {
+      return REPLICA_REJECT;
+    }
+    return ACCEPT;
+  }
+
+  /**
+   * Check whether this goal is violated after removing the given leader replica. It is violated if there is not be enough
+   * leader replicas on that broker for that topic
+   *
+   * @param replicaToBeRemoved replica to be removed
+   * @return {@code true} if the given replica move would violate the optimized goal (i.e. the move is not acceptable),
+   * {@code false} otherwise.
+   */
+  private boolean doesLeaderRemoveViolateOptimizedGoal(Replica replicaToBeRemoved) {
+    if (!replicaToBeRemoved.isLeader()) {
+      // Moving a follower replica does not violate/affect this goal
+      return false;
+    }
+    String topicName = replicaToBeRemoved.topicPartition().topic();
+    if (!_mustHaveTopicLeadersPerBroker.contains(topicName)) {
+      // Moving a replica of a irrelevant topic does not violate/affect this goal
+      return false;
+    }
```

Comment:
This sanity check seems redundant. All the callers of `doesLeaderRemoveViolateOptimizedGoal` already verify that the topicName is in `_mustHaveTopicLeadersPerBroker`.

Am I missing something?

Reply:
Yes, it is redundant in this case